### PR TITLE
[REVIEW] Replace Nulls

### PR DESCRIPTION
**Pull request by aocsa**
This  is a Pull Request. It implements gdf_replace_nulls function.  

gdf_error gdf_replace_nulls(gdf_column* col_out, const gdf_column* new_values_column);

It will take in two gdf_columns.
 * The first one is expected to be a regular gdf_column, the second one has to be a column of the same type as the first, and it has to be of  size one or of the same size as the other column.

 *  case 1: If the second column contains only one value, then this function will replace all nulls in the first column with the value in the second  column.

 * case 2: If the second column is of the same size as the first, then the function will replace all nulls of the first column with the corresponding elements of the  second column 

All the best,

### CODE REVIEW

File: cpp/src/replace/replace.cu [hunk 1]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the apropiate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_kernel_forwarder {
+  template <typename col_type>
+  void operator()(size_t           nrows,
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]:
```suggestion
  void operator()(gdf_size_type           nrows,
```

File: cpp/src/replace/replace.cu [hunk 2]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+                  size_t           new_values_length,
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]:
```suggestion
                  gdf_size_type           new_values_length,
```

File: cpp/src/replace/replace.cu [hunk 3]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
 {
   return find_and_replace_all(col, old_values, new_values);
 }
+
+template <typename Type>
+__global__
+void replace_nulls_with_scalar_kernel(int size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data_scalar)
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]:
```suggestion
void replace_nulls_with_scalar_kernel(gdf_size_type size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data_scalar) 
```

File: cpp/src/replace/replace.cu [hunk 4]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+
+    uint32_t sel_vax = (is_vax_valid >> position) & 1;
+    vax_data[i] = sel_vax? vax_data[i] : *vay_data_scalar;
+  }
+}
+
+
+template <typename Type>
+__global__
+void replace_nulls_with_column_kernel(int size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data)
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]:
```suggestion
void replace_nulls_with_column_kernel(gdf_size_type size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data) 
```

File: cpp/src/replace/replace.cu [hunk 5]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    int index = i / warpSize;
+    uint32_t position = i % warpSize;
+    uint32_t is_vax_valid = vax_valid[index];
+
+    uint32_t sel_vax = (is_vax_valid >> position) & 1;
+    vax_data[i] = sel_vax? vax_data[i] : *vay_data_scalar;
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: This should just use `gdf_is_valid` https://github.com/rapidsai/cudf/blob/branch-0.5/cpp/src/utilities/cudf_utils.h#L21

File: cpp/src/replace/replace.cu [hunk 6]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    int index = i / warpSize;
+    uint32_t position = i % warpSize;
+    uint32_t is_vax_valid = vax_valid[index];
+
+    uint32_t sel_vax = (is_vax_valid >> position) & 1;
+    vax_data[i] = sel_vax? vax_data[i] : vay_data[i];
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: This should use `gdf_is_valid` to check if a value is valid: https://github.com/rapidsai/cudf/blob/branch-0.5/cpp/src/utilities/cudf_utils.h#L21

File: cpp/src/replace/replace.cu [hunk 7]
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+                          col_out->size,
+                          new_values_column->size,
+                          col_out->data,
+                          col_out->valid,
+                          new_values_column->data);
+  return GDF_SUCCESS;
+}
+
+
+//for_each(auto &[data, valid] in zip(col_out->data, col_out->valid) ) {
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: Dead code.

**Comment by eyalroz** [comment]
The explanation at the top of this PR seems to be a piece of text copied from some doxygen comment. "This is a binary function"? No, this is a pull request.

Please rephrase it.

File: cpp/src/replace/replace.cu [hunk 8]
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+    }
+  }
+};
+
+
+gdf_error gdf_replace_nulls(gdf_column* col_out, const gdf_column* new_values_column)
+{
+  GDF_REQUIRE(col_out->dtype == new_values_column->dtype, GDF_DTYPE_MISMATCH);
+  GDF_REQUIRE(new_values_column->size == 1 || new_values_column->size == col_out->size, GDF_COLUMN_SIZE_MISMATCH);
+
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: Needs a few more error checks. Don't want null inputs, and the implementation doesn't support null values in the `new_values_column`.
```suggestion
   GDF_REQUIRE(nullptr != new_values_column->data, GDF_DATASET_EMPTY);
   GDF_REQUIRE(nullptr != col_out->data, GDF_DATASET_EMPTY);
   GDF_REQUIRE(nullptr == new_values_column->valid || 0 == new_values_column->null_count, GDF_VALIDITY_UNSUPPORTED);
```

File: cpp/src/replace/replace.cu [hunk 9]
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
 {
   return find_and_replace_all(col, old_values, new_values);
 }
+
+template <typename Type>
+__global__
+void replace_nulls_with_scalar(gdf_size_type size, Type* out_data, gdf_valid_type * out_valid, const Type *in_data_scalar)
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: Anonymous namespace.

File: cpp/src/replace/replace.cu [hunk 10]
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = gdf_is_valid(out_valid, i)? out_data[i] : *in_data_scalar;
+  }
+}
+
+
+template <typename Type>
+__global__
+void replace_nulls_with_column(gdf_size_type size, Type* out_data, gdf_valid_type* out_valid, const Type *in_data)
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: Anonymous namespace.

File: cpp/src/replace/replace.cu [hunk 11]
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+  }
+}
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the apropiate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_kernel_forwarder {
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: Anonymous namespace.

**Comment by eyalroz** [comment]
I'm sorry to chime in on this at such a late stage, but I have a few reservations I want to share.

1. Requiring that an input column have only two possible lengths is both confusing and surprising. If there's a scalar variant and a columnar variant, make that clearer/explicit, e.g. with a boolean in the signature or with two separate functions. That's IMHO. Alternatively, consider supporting a fixed-size representation of uniform-value columns, in which case both variants are essentially one and the same.
2. Why is the first argument both an input and an output? Is it not "cleaner" to produce a new column that is equal to the first one on non-nulls and to the replacement otherwise?

I realize the design may already be final, so this is just in case it isn't yet.

**Comment by jrhemstad** [comment]
> 1. Requiring that an input column have only two possible lengths is both confusing and surprising. If there's a scalar variant and a columnar variant, make that clearer/explicit, e.g. with a boolean in the signature or with _two separate functions_. That's IMHO. Alternatively, consider supporting a fixed-size representation of uniform-value columns, in which case both variants are essentially one and the same.

Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed). 

>The single-value case specifically can be expressed using the full-column case after applying a fill/replication.

**Comment by aocsa** [comment]
> Please rephrase it.

All right

**Comment by eyalroz** [comment]
> Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).

Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.

I didn't suggest that the two functions be unified using an extra parameter.

**Comment by aocsa** [comment]
> > Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).
> 
> Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.
> 
> I didn't suggest that the two functions be unified using an extra parameter.

In the direction of observation number one of @eyalroz. 

I can change the signature of the `gdf_replace_nulls` adding an extra parameter to specify when the function is used with a scalar and when with a gdf_column.  

In the direction of observation number two of @eyalroz. I agree to change the signature of the  `gdf_replace_nulls`  specifying which one is for input and which one for output. Following the same semantic as `gdf_find_and_replace_all` function. 

Tell me about your thoughts. 
cc @eyalroz @jrhemstad

File: cpp/src/replace/replace.cu (hunk 8 above)
Comment (aocsa) [reply cpp/src/replace/replace.cu]: I agree with these changes. I've already added. Thanks

**Comment by shwina** [comment]
FWIW, the scalar case for `find_and_replace` is handled on the Python side:

https://github.com/rapidsai/cudf/blob/branch-0.6/python/cudf/dataframe/series.py#L777-L781

**Comment by jrhemstad** [comment]
> > Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).
> 
> Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.
> 
> I didn't suggest that the two functions be unified using an extra parameter.

@eyalroz Ah okay, I did misunderstand. However, that approach would require an unnecessary large allocation + memcpy. 

> In the direction of observation number two of @eyalroz. I agree to change the signature of the `gdf_replace_nulls` specifying which one is for input and which one for output. Following the same semantic as `gdf_find_and_replace_all` function.

@aocsa That's fine with me so long as it still supports an in-place operation, i.e., input and output point to the same `gdf_column`.

> FWIW, the scalar case for `find_and_replace` is handled on the Python side:
@shwina we're moving this to libcudf to be more performant and increase the feature completeness of libcudf.

**Comment by aocsa** [comment]
> > > Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).
> > 
> > 
> > Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.

All right! 
> > I didn't suggest that the two functions be unified using an extra parameter.
> 
> @eyalroz Ah okay, I did misunderstand. However, that approach would require an unnecessary large allocation + memcpy.
All right!
> 
> > In the direction of observation number two of @eyalroz. I agree to change the signature of the `gdf_replace_nulls` specifying which one is for input and which one for output. Following the same semantic as `gdf_find_and_replace_all` function.
> 
> @aocsa That's fine with me so long as it still supports an in-place operation, i.e., input and output point to the same `gdf_column`.

I see. So I guess there is not a necessity to add another version. I've just added some basic changes based on the last feedback.  In the last commit, I also rebased the PR, so it is ready to merge. 

All the best.
cc @jrhemstad  

> 
> > FWIW, the scalar case for `find_and_replace` is handled on the Python side:
> > @shwina we're moving this to libcudf to be more performant and increase the feature completeness of libcudf.

File: cpp/src/replace/replace.cu [hunk 12]
```diff
@@ -134,3 +135,96 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
...
+                                            static_cast<const col_type*>(d_new_value)
+                                            );
+      
+    }
+  }
+};
+
+} //end anonymous namespace
+
+gdf_error gdf_replace_nulls(gdf_column* col_out, const gdf_column* new_values_column)
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: `new_values_column` name doesn't agree with name in `functions.h`.

**Comment by jrhemstad** [comment]
rerun tests

File: cpp/src/replace/replace.cu (hunk 12 above)
Comment (aocsa) [reply cpp/src/replace/replace.cu]: done

**Pull request merged**